File: sample/src/main/java/com/github/lany192/box/sample/ui/main/city/CityViewModel.java
package com.github.lany192.box.sample.ui.main.city;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.elvishew.xlog.XLog;
import com.github.lany192.box.sample.bean.Area;
import com.github.lany192.box.sample.http.ApiCallback;
import com.github.lany192.box.sample.http.ApiService;
import com.github.lany192.box.sample.mvvm.BaseViewModel;
import com.hjq.toast.ToastUtils;

import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class CityViewModel extends BaseViewModel implements DefaultLifecycleObserver {
    private final MutableLiveData<List<Area>> items = new MutableLiveData<>();
    private final MutableLiveData<Boolean> loading = new MutableLiveData<>(true);
    private final ApiService apiService;

    @Inject
    public CityViewModel(ApiService apiService) {
        this.apiService = apiService;
        requestCityInfo();
    }

    public MutableLiveData<Boolean> getLoading() {
        return loading;
    }

    public MutableLiveData<List<Area>> getItems() {
        return items;
    }

    public void requestCityInfo() {
        Log.i("TAG:", "请求城市数据接口");
        loading.postValue(true);
        apiService.cityInfo().subscribe(new ApiCallback<List<Area>>() {

            @Override
            public void onSuccess(String msg, List<Area> areas) {
                loading.postValue(false);
                items.postValue(areas);
            }

            @Override
            public void onFailure(String msg, int code) {
                loading.postValue(false);
                ToastUtils.show(msg);
            }
        });
    }


    @Override
    public void onResume(@NonNull LifecycleOwner owner) {
        XLog.i("onResume");
    }

    @Override
    public void onPause(@NonNull LifecycleOwner owner) {
        XLog.i("onPause");
    }
}

File: sample/src/main/java/com/lany192/box/sample/ui/main/index/city/CityViewModel.kt
package com.lany192.box.sample.ui.main.index.city

import androidx.lifecycle.viewModelScope
import com.github.lany192.arch.items.ItemsViewModel
import com.hjq.toast.ToastUtils
import com.lany192.box.sample.repository.BoxRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CityViewModel @Inject constructor(val repository: BoxRepository) : ItemsViewModel() {

    override fun request(refresh: Boolean) {
        viewModelScope.launch {
            repository.getCityList()
                .onStart {
                    log.i("接口开始")
                }.onCompletion {
                    log.i("接口结束")
                }.collect {
                    if (it.code == 0) {
                        if (refresh) {
                            resetItems(it.result)
                            refreshFinish()
                        } else {
                            addItems(it.result)
                            moreLoadFinish()
                        }
                        moreLoadEnd()
                    } else {
                        ToastUtils.show(it.message)
                    }
                }
        }
    }
}